### PR TITLE
Core: Consolidate write.folder-storage.path and write.object-storage.path to write.data.path

### PR DESCRIPTION
**Pull request by flyrain**
Per discussion in #2965, we deprecate both `write.folder-storage.path` and `write.object-storage.path`, and use `write.data.path` instead. 

Created a new method `LocationProviders::dataLocation` to get the data location and to be compatible with deprecated configs.

cc @rdblue @jackye1995 @aokolnychyi @RussellSpitzer @kbendick @karuppayya

### CODE REVIEW

File: core/src/main/java/org/apache/iceberg/TableProperties.java
Comment (kbendick) [inline core/src/main/java/org/apache/iceberg/TableProperties.java:161]: Nit: Do we want to update the comment for the default if not set to reflect anything about the possibility of object storage location provider?

Up to you. The more I think about it, the more I think it just complicates things and that we should just properly document the behavior on the website. But up to you.

File: core/src/main/java/org/apache/iceberg/LocationProviders.java
Comment (kbendick) [inline core/src/main/java/org/apache/iceberg/LocationProviders.java:137]: Nit: Does this need to be `public` or can it be `private`? If it needs to be visible for testing or something, that's ok.

File: core/src/main/java/org/apache/iceberg/LocationProviders.java
Comment (kbendick) [inline core/src/main/java/org/apache/iceberg/LocationProviders.java:149]: While we're falling back through several deprecated options, we might want to list out the intended fallback order / behavior (in a comment up top for example) so we can more easily verify that it happens.

Or better yet, add tests setting different combinations.

We should potentially also address whether or not it's possible for users to have set too many flags and then error out.

Lastly, might want to drop a warning level deprecation log if one of the deprecated flags is non-null (or maybe just info). :)

**Review by kbendick — COMMENTED** [review]
Thanks for handling this @flyrain!

A couple nits and open-ended questions and one more question (for my own understanding).

Clarifying for my own understanding, but if user enables object-storage location provider and _doesnt_ set any of these paths, what happens?

Like in this case, does the data file still get the hash? And if so, where is it placed? `table.location() + "/data" + <hash>`? This might be something to address in another PR (or maybe I just need more coffee).

```java
    table.updateProperties()
        .set(TableProperties.OBJECT_STORE_ENABLED, "true")
        .commit();

    Assert.assertTrue("default data location should be used when object storage path not set",
        table.locationProvider().newDataLocation("file").contains(table.location() + "/data"));
  ```

File: core/src/main/java/org/apache/iceberg/LocationProviders.java
Comment (rdblue) [reply core/src/main/java/org/apache/iceberg/LocationProviders.java:137]: Visible for testing should use protected or package-private.

File: core/src/main/java/org/apache/iceberg/LocationProviders.java
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/LocationProviders.java:140]: This is only correct if the table is using object storage. I think you need to update this logic to select whether to return OBJECT_STORE_PATH or WRITE_FOLDER_STORAGE_LOCATION depending on the location provider selected. Both should fall back to the table location. And the new property should take precedence.

File: core/src/main/java/org/apache/iceberg/TableProperties.java
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/TableProperties.java:139]: I doubt this will ever be removed. It isn't worth breaking on older tables.

File: core/src/main/java/org/apache/iceberg/LocationProviders.java
Comment (flyrain) [reply core/src/main/java/org/apache/iceberg/LocationProviders.java:137]: I make it public because it is used in IcebergSourceBenchmark.java. Maybe we can change `WRITE_FOLDER_STORAGE_LOCATION` to the `WRITE_DATA_PATH` in IcebergSourceBenchmark.java?

**Comment by flyrain** [comment]
> Clarifying for my own understanding, but if user enables object-storage location provider and _doesnt_ set any of these paths, what happens?
> 
> Like in this case, does the data file still get the hash? And if so, where is it placed? `table.location() + "/data" + <hash>`? 

This PR doesn't change the behavior of object storage location provider. It will fall back to `table.location() + "/data"` if no config is set.

File: core/src/main/java/org/apache/iceberg/LocationProviders.java
Comment (flyrain) [reply core/src/main/java/org/apache/iceberg/LocationProviders.java:149]: `testObjectStorageLocationProviderPathResolution` is for that. Will add warn log.

File: core/src/main/java/org/apache/iceberg/TableProperties.java
Comment (flyrain) [reply core/src/main/java/org/apache/iceberg/TableProperties.java]: Make sense, remove the comments.

File: core/src/main/java/org/apache/iceberg/LocationProviders.java
Comment (flyrain) [reply core/src/main/java/org/apache/iceberg/LocationProviders.java]: Agreed. Made the change. This is going to be a little different from #2965. Modified the test case `testObjectStorageLocationProviderPathResolution` and added `testDefaultStorageLocationProviderPathResolution`.

File: core/src/main/java/org/apache/iceberg/LocationProviders.java
Comment (flyrain) [reply core/src/main/java/org/apache/iceberg/LocationProviders.java]: Added the log.

File: core/src/main/java/org/apache/iceberg/LocationProviders.java
Comment (flyrain) [inline core/src/main/java/org/apache/iceberg/LocationProviders.java:147]: We don't have to check if deprecatedProperty is null, the map can take a null.

File: core/src/test/java/org/apache/iceberg/TestLocationProvider.java
Comment (karuppayya) [inline core/src/test/java/org/apache/iceberg/TestLocationProvider.java:238]: nit: should this be `write data should be used when set`?

File: spark3/src/main/java/org/apache/iceberg/spark/actions/BaseSnapshotTableSparkAction.java
Comment (karuppayya) [inline spark3/src/main/java/org/apache/iceberg/spark/actions/BaseSnapshotTableSparkAction.java:170]: When snapshotting a table that has  folder storage location set, should we also remove it in addition to write data location? 
Also should we remove object storage location?

Comment (flyrain) [reply spark3/src/main/java/org/apache/iceberg/spark/actions/BaseSnapshotTableSparkAction.java:172]: It'd be safer to remove both. Added in the new commit.

**Comment by flyrain** [comment]
Hi @jun-he, do we need a similar change in python? table_properties.py contains WRITE_NEW_DATA_LOCATION and WRITE_METADATA_LOCATION as well. I'm glad to add it if the change is needed. Another option is to handle it in a follow-up PR.

File: spark3/src/main/java/org/apache/iceberg/spark/actions/BaseSnapshotTableSparkAction.java
Comment (jackye1995) [reply spark3/src/main/java/org/apache/iceberg/spark/actions/BaseSnapshotTableSparkAction.java:172]: I have a PR #2966 for this change, you can either also port my tests here, or remove this and I will update that PR once this is merged.

**Review by jackye1995 — COMMENTED** [review]
Sorry for the late review, was a bit busy last few days, thanks for working on this!

File: core/src/main/java/org/apache/iceberg/LocationProviders.java
Comment (flyrain) [reply core/src/main/java/org/apache/iceberg/LocationProviders.java:60]: yeah, I was using the formatting tool for the whole file. Will remove it.

File: core/src/main/java/org/apache/iceberg/LocationProviders.java
Comment (flyrain) [reply core/src/main/java/org/apache/iceberg/LocationProviders.java:64]: Will remove it.

File: core/src/main/java/org/apache/iceberg/LocationProviders.java
Comment (flyrain) [reply core/src/main/java/org/apache/iceberg/LocationProviders.java:102]: That's my original logic. I made the change per Ryan's comments https://github.com/apache/iceberg/pull/3094#discussion_r705800767, which makes sense to me.

File: spark3/src/main/java/org/apache/iceberg/spark/actions/BaseSnapshotTableSparkAction.java
Comment (flyrain) [reply spark3/src/main/java/org/apache/iceberg/spark/actions/BaseSnapshotTableSparkAction.java:172]: Hi @jackye1995 , thanks for the information. I've added both `WRITE_FOLDER_STORAGE_LOCATION` and `OBJECT_STORE_PATH`. Can you rebase it in PR #2966 once this got merged?

**Comment by flyrain** [comment]
> nit: we can move L161 to after L144 to avoid deleting the comments and add again here.

I am trying to put these two guys together.
```
  // This only applies to files written after this property is set. Files previously written aren't
  // relocated to reflect this parameter.
  // If not set, defaults to a "data" folder underneath the root path of the table.
  public static final String WRITE_DATA_LOCATION = "write.data.path";

  // This only applies to files written after this property is set. Files previously written aren't
  // relocated to reflect this parameter.
  // If not set, defaults to a "metadata" folder underneath the root path of the table.
  public static final String WRITE_METADATA_LOCATION = "write.metadata.path";
```

File: spark3/src/main/java/org/apache/iceberg/spark/actions/BaseSnapshotTableSparkAction.java
Comment (jackye1995) [reply spark3/src/main/java/org/apache/iceberg/spark/actions/BaseSnapshotTableSparkAction.java:172]: Sure, I can do that

File: core/src/main/java/org/apache/iceberg/LocationProviders.java
Comment (jackye1995) [reply core/src/main/java/org/apache/iceberg/LocationProviders.java:100]: yeah I would agree with that if the change is not in 0.12.0, but unfortunately that PR is a part of the 0.12.0 release. https://github.com/apache/iceberg/compare/apache-iceberg-0.12.0...master

Comment (flyrain) [reply core/src/main/java/org/apache/iceberg/LocationProviders.java:100]: Yeah, the logic of #2845 has been in the release 0.12.0. We'd better to stick to it. I will make the change. cc @rdblue.

Comment (flyrain) [reply core/src/main/java/org/apache/iceberg/LocationProviders.java]: Hi @jackye1995, made the logic to be the same with #2845. The implementation looks not as clean as before, but that's probably the best thing we can do.

File: core/src/main/java/org/apache/iceberg/LocationProviders.java
Comment (jackye1995) [inline core/src/main/java/org/apache/iceberg/LocationProviders.java:77]: nit: is this newline needed?

File: core/src/main/java/org/apache/iceberg/LocationProviders.java
Comment (jackye1995) [inline core/src/main/java/org/apache/iceberg/LocationProviders.java:100]: nit: is this newline needed?

File: core/src/main/java/org/apache/iceberg/LocationProviders.java
Comment (jackye1995) [inline core/src/main/java/org/apache/iceberg/LocationProviders.java:145]: I think we can use `String deprecatedProperty = isObjectStorage ? TableProperties.OBJECT_STORE_PATH : TableProperties.WRITE_FOLDER_STORAGE_LOCATION`, then `deprecatedProperty` is always not null, and you don't need to do null check for the warning, and we can avoid having that warning helper method.

Comment (flyrain) [reply core/src/main/java/org/apache/iceberg/LocationProviders.java]: Made the change. Thanks for the suggestion.

File: site/docs/aws.md
Comment (jackye1995) [inline site/docs/aws.md:377]: Given we have a simple path resolution strategy, I think we can just put these listing in a single sentence. Also add 2 warning blocks describing the legacy behaviors:
- before 0.12.0, `write.object-storage.path` must be set
- at 0.12.0, `write.object-storage.path` then `write.folder-storage.path` then `<tableLocation>/data`

**Review by jackye1995 — APPROVED** [review]
Looks good to me! I have a small comment around documentation, apart from that I don't have additional feedbacks. Let's see if anyone else has additional concern, if not I will merge it after some time.

File: site/docs/aws.md
Comment (flyrain) [reply site/docs/aws.md]: Thanks Jack! Made the change for the doc.

**Comment by flyrain** [comment]
Thanks all for the review. Is it ready for merging?

File: core/src/main/java/org/apache/iceberg/LocationProviders.java
Comment (kbendick) [inline core/src/main/java/org/apache/iceberg/LocationProviders.java:146]: Nit / non-blocking: Naming this `isObjectStorageLocationProvider` or something to indicate that it's not just object storage, but specifically the usage of object storage location provider.

However, since I just realized this is already in `LocationProviders.java`, I think this is probably fine and no need to change (especially as it's documented in the java doc).

**Review by kbendick — APPROVED** [review]
Overall this looks good to me. Thank you for doing this @flyrain!

**Comment by jun-he** [comment]
> Hi @jun-he, do we need a similar change in python? table_properties.py contains WRITE_NEW_DATA_LOCATION and WRITE_METADATA_LOCATION as well. I'm glad to add it if the change is needed. Another option is to handle it in a follow-up PR.
@flyrain thanks for pinging. Yep, we need to have the similar changes. But I think it is fine not to add it in the python_legacy. Instead, we can add it to the new python library. I am currently working on its layout with a high level design. Once that is done, we will start the implementation and then add it to python. I will create an issue for us to track it.
Additionally, wondering if we still need to support those deprecated fields for backward compatibility?

File: core/src/main/java/org/apache/iceberg/LocationProviders.java
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/LocationProviders.java:174]: This looks way too complicated to be worth it. Let's just duplicate the logic and minimize the changes in each location provider.

File: core/src/main/java/org/apache/iceberg/LocationProviders.java
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/LocationProviders.java:170]: No need for warnings. This is not important enough to nag users about.

Comment (flyrain) [reply core/src/main/java/org/apache/iceberg/LocationProviders.java]: Removed

File: core/src/main/java/org/apache/iceberg/LocationProviders.java
Comment (flyrain) [reply core/src/main/java/org/apache/iceberg/LocationProviders.java]: Made the change.

**Comment by rdblue** [comment]
Looks like the change that defaulted object storage tables to the folder-storage path has not been released yet: https://github.com/apache/iceberg/commit/aef12c0e702ef8c306826581893963e3dc5f6ba3

@flyrain and @jackye1995, how about simplifying this by only using `write.data.path` and removing the use of `write.folder-storage.path` from the object store location provider? That simplifies the documentation and makes it easier to understand.

**Comment by flyrain** [comment]
Hi @rdblue, I like your idea, but as @jackye1995 mentioned, the logic to use write.folder-storage.path in the object store location provider was introduced by #2845, which has been released in 0.12.0. #2965 is a PR to just change the constant name.

**Comment by jackye1995** [comment]
@rdblue @flyrain 

I think if 1.0.0 is the next release, what @rdblue proposes is definitely the right approach to go, and we have also a way out for users of 0.12.0, which is great.

If we are still thinking about 0.13.0, we need to just be more cautious, because these config keys will be there all the time, I think it's better to keep the resolution strategy. We decided to merge #2845 mostly because services try to hide the table's root location, but the object storage mode forces user to configure a location that the user might not know. Chaining the fallback config key seems to be the only way out for the 2 use cases described here: https://github.com/apache/iceberg/pull/2845#issuecomment-897900861

**Comment by rdblue** [comment]
I don't think that we shouldn't make a breaking behavior change at 1.0. It isn't important enough to clean this up to warrant a breaking change. Let's just continue with the released behavior as a fallback.

**Review by rdblue — APPROVED** [review]

**Comment by rdblue** [comment]
Thanks, @flyrain!

**Comment by flyrain** [comment]
Thanks all for the review and commit. @rdblue @jackye1995 @kbendick @karuppayya @jun-he. 
Hi @anuragmantri, fyi, this will affect the relative path design a bit.

**Pull request merged**
